server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -985,36 +1020,116 @@ public void endVerification(String seed) {
 
     // Tracks the latest known repository generation in a best-effort way to detect inconsistent listing of root level index-N blobs
     // and concurrent modifications.
-    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.EMPTY_REPO_GEN);
+    private final AtomicLong latestKnownRepoGen = new AtomicLong(RepositoryData.UNKNOWN_REPO_GEN);
 
     @Override
     public void getRepositoryData(ActionListener<RepositoryData> listener) {
-        ActionListener.completeWith(listener, () -> {
-            // Retry loading RepositoryData in a loop in case we run into concurrent modifications of the repository.
-            while (true) {
+        if (latestKnownRepoGen.get() == RepositoryData.CORRUPTED_REPO_GEN) {
+            listener.onFailure(corruptedStateException(null));
+            return;
+        }
+        if (bestEffortConsistency == false && latestKnownRepoGen.get() == RepositoryData.UNKNOWN_REPO_GEN) {
+            if (clusterService.state().nodes().isLocalNodeElectedMaster() == false) {
+                // Non-master node does not load RepositoryData before master has at least loaded it once and thus initialized in the
+                // cluster state
+                listener.onFailure(new RepositoryException(metadata.name(), "Repository not initialized yet"));
+            } else {
+                // After initializing the repo in the cluster state we retry loading the latest RepositoryData on the generic pool
+                initializeRepoInClusterState(ActionListener.wrap(v ->
```

Comment:
I don't think we should initialize this here (as it can be based on an inconsistent read). Instead I prefer we add this after a successful write repo operation and load leniently until that has happened.

Reply:
Thanks, that's a sweet idea I think. That way we force at least one more listing before writing the new repo generation and are even safer in a freaky scenario where the repo was just modified and then mounted for writing :) Will adjust accordingly.

Reply:
Removed the initialization logic in the read path completely now and simply made `BlobStoreRepository` use the old style listing fallback as long as the cluster state contains `RepositoryData#UNKNOWN_REPO_GEN` in the `RepositoryMetaData`. That also solves all the BwC issues IMO, since if we fall over from an older master, the cluster state will container `RepositoryData#UNKNOWN_REPO_GEN` and if we're already on a new version master then the contents of the cluster state will be correct if that master has ever written a new `index-N` or will be `RepositoryData#UNKNOWN_REPO_GEN` otherwise.